ethernet: phy: extend phy interface with MMD read and write

go2sh opened the pull request:
Extend the phy interface to read and write MMD device registers. To be used in conjunction with clause 45 mdio requests.

go2sh committed:
net: ethernet: extend phy api with MMD capable read and write

The phy device api is exetended with read and write functions
for phys, which support MDIO manageable devices. These phys have
16bit register spaces for each of the 32 MMD devices.

Signed-off-by: Christoph Seitz <christoph.seitz@infineon.com>

carlescufi commented:
Thanks for the PR! Please take a look at [our commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) to find out how to format your commit messages, and at [our contribution workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) to understand how to update your Pull Request.

pdgendt reviewed (dismissed):
What problem are you trying to solve? Provide samples or tests, and at minimum one supported driver.

Have you seen #77817 where similar changes were requested, but can be solved in device tree using child nodes.

go2sh commented:
I'll add test and implements this in one driver. 

The general topic is that modern phys have a feature called mdio manageable devices, which describe sub devices inside a phy, like pcs and pma. The address of a register for that consists of the mmd devices address and the register address. So the adress space is 32x larger. This function would for most devices directly map on the corresponding mdio_read_c45 function or on the emulation mechanism for c45 accesses based on clause 22. 

For a phy and underlying interface agnostic interface, these functions are required.

decsny commented:
@manuargue added mdio c45 api to zephyr, let's get his opinion about this

go2sh commented:
One more example would also be the xpcs phy from synopsis. There the phy can also be part of the soc and the register access is memory mapped. 

Also the MMD register spaces are standardised within the ieee. So a follow-up pr could contain the register map.

go2sh committed:
drivers: adin2111: Implement phy device api for adin2111

Extend the phy api for the adin2111 with the phy_read_device
and phy_write_device functions based on the already
available c45 functions.

Signed-off-by: Christoph Seitz <christoph.seitz@infineon.com>

go2sh committed:
tests: drivers: Add tests for new phy apis

Add test for phy new apis phy_read_device and phy_write_device.

Signed-off-by: Christoph Seitz <christoph.seitz@infineon.com>

go2sh commented:
Implemented the feedback:
* Commit messages updated
* Added tests for the API
* Extended one driver to use the API, where the functions were already available.

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -118,6 +119,14 @@ __subsystem struct ethphy_driver_api {
 		     uint32_t data);
+
+	/** Read PHY device register */
+	int (*read_device)(const struct device *dev, uint8_t dev_addr, uint16_t reg_addr,
```
```suggestion
	int (*read_mmd)(const struct device *dev, uint8_t dev_addr, uint16_t reg_addr,
```
and `write_mmd` likewise, to align with Linux PHY APIs

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -118,6 +119,14 @@ __subsystem struct ethphy_driver_api {
+
+	/** Read PHY device register */
+	int (*read_device)(const struct device *dev, uint8_t dev_addr, uint16_t reg_addr,
+			   uint16_t *data);
```
```suggestion
			   uint32_t *data);
```
and same for the write function, to be consistent with existing `phy_read|write` functions

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -232,6 +241,67 @@ static inline int phy_write(const struct device *dev, uint16_t reg_addr,
 }
 
+/**
+ * @brief      Read PHY registers
+ *
+ * This routine provides a generic interface to read from a PHY register.
```
```suggestion
 * @brief      Read PHY register from an MMD
 *
 * This routine provides a generic interface for reading a register from an MMD
 * on a given PHY.
```

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -232,6 +241,67 @@ static inline int phy_write(const struct device *dev, uint16_t reg_addr,
+ * This routine provides a generic interface to read from a PHY register.
+ *
+ * @param[in]  dev       PHY device structure
+ * @param[in]  dev_addr  MMD device address
```
```suggestion
 * @param[in]  dev_addr  MMD address
```
D=device

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -232,6 +241,67 @@ static inline int phy_write(const struct device *dev, uint16_t reg_addr,
+}
+
+/**
+ * @brief      Write PHY register
+ *
+ * This routine provides a generic interface to write to a PHY register.
```
```suggestion
 * @brief      Write PHY register on an MMD
 *
 * This routine provides a generic interface for writing a register on an MMD
 * on a given PHY.
```

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -232,6 +241,67 @@ static inline int phy_write(const struct device *dev, uint16_t reg_addr,
+ * This routine provides a generic interface to write to a PHY register.
+ *
+ * @param[in]  dev       PHY device structure
+ * @param[in]  dev_addr  MMD device address
```
```suggestion
 * @param[in]  dev_addr  MMD address
```

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -232,6 +241,67 @@ static inline int phy_write(const struct device *dev, uint16_t reg_addr,
+		return -ENOSYS;
+	}
+
+	if (dev_addr > 0x1F) {
```
```suggestion
	if (dev_addr > MII_MMD_ACR_DEVAD_MASK) {
```
from `<zephyr/net/mii.h>`

go2sh commented:
@manuargue Addressed your feedback, please review. :)

decsny commented:
> What problem are you trying to solve? Provide samples or tests, and at minimum one supported driver.
> 
> Have you seen #77817 where similar changes were requested, but can be solved in device tree using child nodes.

To be clear, that was different, the other PR was basically trying to add a parameter to the existing API for port address, which was wrong becuase:
* redundant since that is already abstracted by the zephyr phy device, the port address is given in DT and can therefore be used by the driver without client needing to know or specify it in code. 
* breaking change the way it was done

This PR is adding an new API that adds support to address the MMD devices introduced to ethernet phys in 802.3 clause 45, which I think is valid, and isn't being done in a way that would break the old API

GeorgeCGV commented on drivers/ethernet/phy/phy_adin2111.c:
```diff
@@ -214,6 +214,36 @@ static int phy_adin2111_reg_write(const struct device *dev, uint16_t reg_addr,
+	const struct phy_adin2111_config *cfg = dev->config;
+	int ret;
+
+	mdio_bus_enable(cfg->mdio);
```
Be careful with `mdio_bus_enable|disable` for ADIN. They are used to lock|unlock the MAC.

go2sh replied:
This is a copy of the c22 phy_read/write Implementation for c44. If this is wrong, then the c22 Implementation is also wrong.

go2sh replied:
The problem why a new function is needed due to the conversion of uint16 and uint32 in the phy api.

GeorgeCGV replied:
Won't it be better to use `uint16_t data` for the mmd?

go2sh replied:
@manuargue commented to use 32bit to be consistent with the old phy read write api.

GeorgeCGV replied:
Maybe it is better not to have mmd api but correct existing c22 to use `uint16_t`? Clause 22 registers are 16 bits wide.

go2sh replied:
The mmd api is needed as you need three addresses to access a register (phy, device, register). 

The general question is if 32bit was the right choice for the old read/writw api to begin  with. Mmds are also only 16bit. My intention to add this was, that I have a device, where the phy ist part of the soc and the api is very handy, to hide the why to access the phy.

I'm fine with removing this change entirety. I was ask to implement a reference, I have a different driver, which a want to publish.

manuargue replied:
> Won't it be better to use uint16_t data for the mmd in the API?

My suggestion to use `uint32_t` was to be consistent with the existent phy reg API, although I agree that it should be better to switch both APIs to `uint16_t`. Doing so will need to have all drivers consuming the API updated accordingly and it should be documented in the migration notes as it's a breaking change.

> The general question is if 32bit was the right choice for the old read/writw api to begin with. Mmds are also only 16bit.

The only reason I can think of it was added as `uint32_t`, is if it was supposed to hold also the MMD in the upper 16-bits, but this is just speculation.

go2sh replied:
> The only reason I can think of it was added as uint32_t, is if it was supposed to hold also the MMD in the upper 16-bits, but this is just speculation.

Now the API allows also a different mapping between mdio and phy registers, where 32bits can be used.

All the implantations if seen in zephyr are wrappers around the mdio functions.

manuargue replied:
> Now the API allows also a different mapping between mdio and phy registers, where 32bits can be used.

Right, I was thinking only in the context of 802.3 protocol, but in fact the `phy_read|write` APIs could be implemented for other management interfaces using 32-bit data (like an in-band host interface for example).

For the MMD API, since it seems to be tied to 802.3 C45, then it makes sense to keep it 16-bit as it was originally proposed.

decsny commented on include/zephyr/net/phy.h:
```diff
@@ -118,6 +120,14 @@ __subsystem struct ethphy_driver_api {
 		     uint32_t data);
+
+	/** Read PHY register on an MMD */
+	int (*read_mmd)(const struct device *dev, uint8_t dev_addr, uint16_t reg_addr,
```
would it be worth it to make these apis conditional on a kconfig ?

go2sh replied:
Think it's fine to leave this out. There is no additional cost for this in current projects, except the two function pointers for the new api. But I can also add this. New drivers need to adapt this anyhow.

decsny commented:
@go2sh any plan to continue with the PR?

go2sh commented:
Hi, exect your comment, all seams to be resolved. The general problem is, that there are phys that use mmds and phys that use the old mii space. Some can do both.

decsny commented:
@pdgendt can you revisit

decsny commented:
@go2sh PR needs rebase

decsny commented:
@go2sh reminder of this PR
